RecoBTag/FeatureTools/plugins/HiggsInteractionNetTagInfoProducer.cc

```diff
@@ -0,0 +1,353 @@
+    const auto *packed_cand = dynamic_cast<const pat::PackedCandidate *>(&(*cand));
+    const auto *reco_cand = dynamic_cast<const reco::PFCandidate *>(&(*cand));
+
+    if (packed_cand && !packed_cand->hasTrackDetails()) {
+      continue;
+    } else if (reco_cand && !useTrackProperties(reco_cand)) {
+      continue;
+    }
+
+    auto candP4 = cand->p4();
+    if (packed_cand) {
+      fts.fill("cpf_quality", packed_cand->pseudoTrack().qualityMask());
+
+      // impact parameters
+      fts.fill("cpf_dz", catch_infs(packed_cand->dz()));
+      fts.fill("cpf_dzsig", catch_infs_and_bound(packed_cand->dz() / packed_cand->dzError(), 0, -2000, 2000));
+      fts.fill("cpf_dxy", catch_infs(packed_cand->dxy()));
+      fts.fill("cpf_dxysig", catch_infs_and_bound(packed_cand->dxy() / packed_cand->dxyError(), 0, -2000, 2000));
+    } else if (reco_cand) {
+      fts.fill("cpf_quality", quality_from_pfcand(*reco_cand));
+
+      // impact parameters
+      const auto *trk = reco_cand->bestTrack();
+      float dz = trk->dz(pv_->position());
+      float dxy = trk->dxy(pv_->position());
+      fts.fill("cpf_dz", catch_infs(dz));
+      fts.fill("cpf_dzsig", catch_infs_and_bound(dz / trk->dzError(), 0, -2000, 2000));
+      fts.fill("cpf_dxy", catch_infs(dxy));
+      fts.fill("cpf_dxysig", catch_infs_and_bound(dxy / trk->dxyError(), 0, -2000, 2000));
+    }
+
+    // basic kinematics
+    fts.fill("cpf_phirel", reco::deltaPhi(candP4, jet));
+    fts.fill("cpf_etarel", etasign * (candP4.eta() - jet.eta()));
+    fts.fill("cpf_deltaR", reco::deltaR(candP4, jet));
+
+    fts.fill("cpf_ptrel", catch_infs(candP4.pt() / jet.pt(), -99));
+    fts.fill("cpf_erel", catch_infs(candP4.energy() / jet.energy(), -99));
+
+    float drminpfcandsv = btagbtvdeep::mindrsvpfcand(*svs_, &(*cand), jet_radius_);
+    fts.fill("cpf_drminsv", catch_infs_and_bound(drminpfcandsv, 0, -1. * jet_radius_, 0, -1. * jet_radius_));
+
+    // subjets
+    auto subjets = patJet->subjets();
+    std::sort(subjets.begin(), subjets.end(), [](const edm::Ptr<pat::Jet> &p1, const edm::Ptr<pat::Jet> &p2) {
+      return p1->pt() > p2->pt();
+    });  // sort by pt
+    fts.fill("cpf_drsubjet1", !subjets.empty() ? reco::deltaR(*cand, *subjets.at(0)) : -1);
+    fts.fill("cpf_drsubjet2", subjets.size() > 1 ? reco::deltaR(*cand, *subjets.at(1)) : -1);
+
+    const reco::Track *trk = nullptr;
+    if (packed_cand && packed_cand->hasTrackDetails()) {
+      trk = &(packed_cand->pseudoTrack());
+    } else if (reco_cand && useTrackProperties(reco_cand)) {
+      trk = reco_cand->bestTrack();
+    }
+    TrackInfoBuilder trkinfo(track_builder_);
+    trkinfo.buildTrackInfo(&(*cand), jet_dir, jet_ref_track_dir, *pv_);
+    fts.fill("cpf_btagEtaRel", catch_infs_and_bound(trkinfo.getTrackEtaRel(), 0, -5, 15));
+    fts.fill("cpf_btagPtRatio", catch_infs_and_bound(trkinfo.getTrackPtRatio(), 0, -1, 10));
+    fts.fill("cpf_btagPParRatio", catch_infs_and_bound(trkinfo.getTrackPParRatio(), 0, -10, 100));
```

Comment:
The choice of negative values seems a bit arbitrary.
are the negative bounds even possible?

Reply:
These choices were made just to match the original ntupler used for training (https://github.com/DeepDoubleB/DNNTuplesAK8/blob/opendata_80X/BTagHelpers/src/TrackInfoBuilder.cc#L56-L70). 
This results in 100% agreement for the tagger output. 

Changing these (here and elsewhere) will have some (possibly small) effect on the tagger output and possibly break the 100% agreement.

I can check how far we are from 100% and see if we can live with it (i.e. if tagger performance won't be greatly affected).

Reply:
my comment implied that all the "Rel" are with respect to the jet axis; if the cands are from the jet, how can any of the values be negative?
Hence, the negative values as lower bounds are just random.